src/engine/MaxConstraint.cpp

```diff
@@ -121,64 +123,83 @@ void MaxConstraint::registerAsWatcher( ITableau *tableau )
 void MaxConstraint::unregisterAsWatcher( ITableau *tableau )
 {
     for ( unsigned element : _initialElements )
+    {
         tableau->unregisterToWatchVariable( this, element );
+        if ( _elementToAux.exists( element ) )
+            tableau->unregisterToWatchVariable( this, _elementToAux[element] );
+    }
 
     if ( !_initialElements.exists( _f ) )
         tableau->unregisterToWatchVariable( this, _f );
 }
 
 void MaxConstraint::notifyVariableValue( unsigned variable, double value )
 {
-    if ( ( _elements.exists( _f ) || variable != _f ) &&
-         ( !maxIndexSet() || !_assignment.exists( getMaxIndex() ) || _assignment.get( getMaxIndex() ) < value ) )
-    {
-        setMaxIndex( variable );
-    }
     _assignment[variable] = value;
 }
 
 void MaxConstraint::notifyLowerBound( unsigned variable, double value )
 {
     if ( _statistics )
-        _statistics->incLongAttribute( Statistics::NUM_BOUND_NOTIFICATIONS_TO_PL_CONSTRAINTS );
+        _statistics->incLongAttribute
+            ( Statistics::NUM_BOUND_NOTIFICATIONS_TO_PL_CONSTRAINTS );
 
-    if ( existsLowerBound( variable ) && !FloatUtils::gt( value, getLowerBound( variable ) ) )
+    if ( existsLowerBound( variable ) &&
+         !FloatUtils::gt( value, getLowerBound( variable ) ) )
         return;
 
     setLowerBound( variable, value );
 
-    bool maxErased = false;
-
-    if ( _elements.exists( variable ) && FloatUtils::gt( value, _maxLowerBound ) )
+    /*
+      See if we can eliminate any cases.
+    */
+    if ( _auxToElement.exists( variable ) )
     {
-        _maxLowerBound = value;
-        List<unsigned> toRemove;
-        for ( auto element : _elements )
+        // The case that this variable is an aux variable.
+        // We can eliminate the corresponding case if the aux variable is
+        // positive.
+        if ( FloatUtils::isPositive( value ) )
+            eliminateCase( _auxToElement[variable] );
+    }
+    else if ( variable == _f || _elements.exists( variable ) )
+    {
+        // If the variable is either in _element or _f. The only case that this is
+        // going to eliminate case is that the new lowerBound is greater than the
+        // _maxLowerBound.
+        // We update _maxLowerBound, and go through each element to see if it
+        // can be eliminated.
+        // We also update the _haveFeasibleEliminatedCase flag.
```

Comment:
This comment can be streamlined:

// When the variable is either in _element or _f, a case is eliminated,
//  when the new lowerBound is greater than the _maxLowerBound.

The comment should not be stating things that are obvious when reading the code. Lines 169-170 would be somewhat appropriate if this was a method description, but here the code follows and the reader can see what is done.

Reply:
Done